File: uva/10927 Bright lights/lights.py
from fractions import gcd

class point:
    def __init__(self, x: int, y: int, height: int):
        self.x = x
        self.y = y
        if y != 0:
            g = gcd(x, y)
            self.xy = (x // g, y // g)
        elif x > 0:
            self.xy = (1,)
        else:
            self.xy = (-1,)
        self.height = height
    
    def __lt__(self, other):
        if self.xy == other.xy:
            return self.x ** 2 + self.y ** 2 < other.x ** 2 + other.y ** 2
        if self.xy == (1,) or other.xy == (-1,):
            return False
        elif self.xy == (-1,) or other.xy == (1,):
            return True
        x1, y1 = self.xy
        x2, y2 = other.xy
        return x1 * y2 < x2 * y1

caseNum = 1

n = int(input())
while n > 0:
    poles = []
    for _ in range(n):
        x, y, z = map(int, input().split())
        # x, y, z = random.randint(-100000, 100000), random.randint(0, 10000), random.randint(0, 10000)
        poles.append(point(x, y, z))
    
    n = int(input())
    continue
    poles = sorted(poles)
    
    last = poles[0]
    maxH = last.height
    ans = []
    for pole in poles[1:]:
        if pole.xy == last.xy:
            # same angle as the last pole; potential for overlap
            if pole.height <= maxH:
                ans.append((pole.x, pole.y))
            else:
                maxH = pole.height
        else:
            maxH = pole.height
        last = pole
    
    print("Data set {}:".format(caseNum))
    if len(ans) == 0:
        print("All the lights are visible.")
    else:
        print("Some lights are not visible:")
        ans = sorted(ans)
        for p in ans:
            if p == ans[-1]:
                print("x = {}, y = {}.".format(*p))
            else:
                print("x = {}, y = {};".format(*p))
    caseNum += 1
    n = int(input())

File: uva/10927 Bright lights/lights.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;

int caseNum = 1;
class Angle {
public:
    ll x, y;
    Angle(ll _x, ll _y) {
        ll g = __gcd(_x, _y);
        g = abs(g);
        // if (_y == 0 and _x < 0) {
            // x = -1;
        // } else {
            x = _x / g;
        // }
        y = _y / g;
    }
    bool operator==(Angle other) {
        return x == other.x and y == other.y;
    }
    bool isInf() {
        return x > 0 and y == 0;
    }
    bool isNegInf() {
        return x < 0 and y == 0;
    }
    bool operator<(Angle other) {
        return x * other.y < other.x * y;
    }
};

class point {
public:
    ll x, y, height;
    Angle angle;
    point(ll _x, ll _y, ll _height) : x(_x), y(_y), height(_height), angle(_x, _y) {
        
    }
    
    ll abs() {
        return x * x + y * y;
    }
    
    const bool operator<(point other) {
        
        if (angle == other.angle) {
            return abs() < other.abs();
        }
        if (angle.isInf() or other.angle.isNegInf()) {
            return false;
        } else if (angle.isNegInf() or other.angle.isInf()) {
            return true;
        }
        return angle < other.angle;
    }
};

int main() {
    int n;
    cin >> n;
    while (n > 0) {
        vector<point> poles;
        for (int i = 0; i < n; i ++) {
            int x, y, z;
            cin >> x >> y >> z;
            poles.push_back(point(x, y, z));
        }
        
        sort(poles.begin(), poles.end());
        // if (caseNum == 91) {
        //     for (point pole: poles)
        //     cout << pole.x << " " << pole.y << " " << pole.height << endl;
        // }
        
        point last = poles[0];
        
        ll maxH = last.height;
        
        vector<pair<int, int>> ans;
        
        bool first = true;
        for (point pole: poles) {
            if (first) {
                first = false;
                continue;
            }
            if (pole.angle == last.angle) {
                // same angle as the last pole; potential for overlap
                if (pole.height <= maxH) {
                    ans.push_back({pole.x, pole.y});
                } else {
                    maxH = pole.height;
                }
            } else {
                maxH = pole.height;
            }
            last = pole;
        }
        cout << "Data set " << caseNum ++ << ":" << endl;
        if (ans.size() == 0) {
            cout << "All the lights are visible." << endl;
        } else {
            cout << "Some lights are not visible:" << endl;
            sort(ans.begin(), ans.end());
            for (auto p: ans) {
                if (p == ans[ans.size() - 1]) {
                    cout << "x = " << p.first << ", y = " << p.second << "." << endl;
                } else {
                    cout << "x = " << p.first << ", y = " << p.second << ";" << endl;
                }
            }
        }
        
        cin >> n;
    }
}
